x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/build_exceptions_query.ts

```diff
@@ -0,0 +1,151 @@
+  if (language === 'lucene') {
+    return value.toUpperCase();
+  }
+
+  return value;
+};
+
+export const operatorBuilder = (operator: ListOperator, language: string): string => {
+  return operator === 'excluded'
+    ? ` ${getLanguageBooleanOperator(language, 'and')} `
+    : ` ${getLanguageBooleanOperator(language, 'and')} ${getLanguageBooleanOperator(
+        language,
+        'not'
+      )} `;
+};
+
+export const buildExists = ({
+  operator,
+  field,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  language: string;
+}): string => {
+  switch (language) {
+    case 'kuery':
+      return `${operatorBuilder(operator, language)}${field}:*`;
+    case 'lucene':
+      return `${operatorBuilder(operator, language)}_exists_${field}`;
+    default:
+      return '';
+  }
+};
+
+export const buildMatch = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  const value = values[0].name;
+  return `${operatorBuilder(operator, language)}${field}:${value}`;
+};
+
+export const buildMatchAll = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  if (values.length === 1) {
+    return buildMatch({ operator, field, values, language });
+  } else {
+    const matchAllValues = values.map(value => {
+      return value.name;
+    });
+
+    return `${operatorBuilder(operator, language)}${field}:(${matchAllValues.join(
+      ` ${getLanguageBooleanOperator(language, 'or')} `
+    )})`;
+  }
+};
+
+export const evaluateValues = ({ list, language }: { list: List; language: string }): string => {
+  const { values_operator: operator, values_type: type, field, values } = list;
+  switch (type) {
+    case 'exists':
+      return buildExists({ operator, field, language });
+    case 'match':
+      return buildMatch({ operator, field, values: values!, language });
+    case 'match_all':
+      return buildMatchAll({ operator, field, values: values!, language });
+    default:
+      return '';
+  }
+};
+
+export const buildExceptions = ({
+  query,
+  lists,
+  language,
+  accumulator = [],
+}: {
+  query: string;
+  lists: List[];
+  language: string;
+  accumulator?: string[];
+}): string[] => {
+  return lists.reduce((acc, item) => {
+    const exception: string[] = [query];
+    const { and, ...exceptionDetails } = { ...item };
+
+    exception.push(evaluateValues({ list: exceptionDetails, language }));
+
+    if (and && !isEmpty(and)) {
+      and.forEach(i => {
+        exception.push(evaluateValues({ list: i, language }));
+      });
+    }
+
+    acc.push(`(${exception.join('')})`);
+    return acc;
+  }, accumulator);
+};
+
+export const buildQueryExceptions = ({
+  query,
+  language,
+  lists,
+}: {
+  query: string;
+  language: string;
+  lists?: RuleAlertParams['lists'];
```

Comment:
You want to avoid at all possible costs any optional parameters such as this one:

```ts
lists?: RuleAlertParams['lists'];
```

It's better to actually force people to push down `undefined`. Looking at your types you don't need the question mark at all except for a unit test that you can delete. I would just remove that test and the question mark here. Makes for less code.

Reply:
I think I originally didn't want to have people have to put down `undefined`, but I see your point. I added more tests to account for that too. I'll be sure to try to avoid this pattern in the future.